lib/smack/DSAWrapper.cpp

```diff
@@ -22,185 +24,94 @@ RegisterPass<DSAWrapper>
-  else if (const Argument *A = dyn_cast<Argument>(V))
-    return TD->getDSGraph(*A->getParent());
-  else if (const BasicBlock *BB = dyn_cast<BasicBlock>(V))
-    return TD->getDSGraph(*BB->getParent());
-  else if (isa<GlobalValue>(V))
-    return TD->getGlobalsGraph();
-
-  // XXX I know this looks bad, but it works for now
-  for (auto U : V->users()) {
-    return getGraphForValue(U);
-  }
-
-  llvm_unreachable("Unexpected value.");
 }
 
-int DSAWrapper::getOffset(const MemoryLocation *l) {
-  const DSGraph::ScalarMapTy &S = getGraphForValue(l->Ptr)->getScalarMap();
-  DSGraph::ScalarMapTy::const_iterator I = S.find((const Value *)l->Ptr);
-  if (I == S.end())
-    return 0;
-  if (I->second.getNode() && I->second.getNode()->isCollapsedNode())
-    return -1;
-  unsigned offset = I->second.getOffset();
-  assert(offset <= INT_MAX && "Cannot handle large offsets");
-  return (int)offset;
-}
-
-bool DSAWrapper::isMemcpyd(const llvm::DSNode *n) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  const llvm::DSNode *nn = eqs.getLeaderValue(n);
-  for (unsigned i = 0; i < memcpys.size(); i++)
-    if (memcpys[i] == nn)
-      return true;
-  return false;
+void DSAWrapper::collectMemCpyds(llvm::Module &M) {
+  for (auto& f : M) {
+    for (inst_iterator I = inst_begin(&f), E = inst_end(&f); I != E; ++I) {
+      if (MemCpyInst* memcpyInst = dyn_cast<MemCpyInst>(&*I)) {
+        auto srcNode = getNode(memcpyInst->getSource());
+        auto destNode = getNode(memcpyInst->getDest());
+        if (srcNode)
+          memCpyds.insert(srcNode);
+        if (destNode)
+          memCpyds.insert(destNode);
+      }
+      if (MemSetInst* memsetInst = dyn_cast<MemSetInst>(&*I)) {
+        auto destNode = getNode(memsetInst->getDest());
+        if (destNode)
+          memCpyds.insert(destNode);
+      }
+    }
+  }
 }
 
-bool DSAWrapper::isStaticInitd(const llvm::DSNode *n) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  const llvm::DSNode *nn = eqs.getLeaderValue(n);
-  for (unsigned i = 0; i < staticInits.size(); i++)
-    if (staticInits[i] == nn)
-      return true;
-  return false;
+void DSAWrapper::countGlobalRefs() {
+  for (auto& g : DG->globals()) {
+    auto& cellRef = g.second;
+    auto* node = cellRef->getNode();
+    if (node) {
+      if (!globalRefCount.count(node))
+        globalRefCount[node] = 1;
+      else
+        globalRefCount[node]++;
+    }
+  }
 }
 
-bool DSAWrapper::isFieldDisjoint(const llvm::Value *V,
-                                 const llvm::Function *F) {
-  return TS->isFieldDisjoint(V, F);
+bool DSAWrapper::isStaticInitd(const sea_dsa::Node *n) {
+  return staticInits.count(n) > 0;
 }
 
-bool DSAWrapper::isFieldDisjoint(const GlobalValue *V, unsigned offset) {
-  return TS->isFieldDisjoint(V, offset);
+bool DSAWrapper::isMemCpyd(const sea_dsa::Node *n) {
```

Comment:
This should be renamed since it is referring to both memcpy and memset.
Also, what about memmove? Also, are there any others?

Reply:
This is still outstanding.